snuba/datasets/events.py

```diff
@@ -260,6 +260,14 @@ def __init__(self) -> None:
             partition_by="(toMonday(timestamp), if(equals(retention_days, 30), 30, 90))",
             version_column="deleted",
             sample_expr=sample_expr,
+            required_deletion_columns=[
+                "event_id",
+                "project_id",
+                "group_id",
+                "timestamp",
+                "deleted",
+                "retention_days",
+            ],
```

Comment:
I think it's important to be able to clearly articulate why columns end up in this set. Ideally this set wouldn't even need to be explicitly defined like this in the future, since I think most of these are derived from other aspects of the schema definition.

My interpretation:

- `project_id` and `event_id` are essentially a composite primary key, so both fields need to be in here, otherwise the tombstoning strategy wouldn't work. These columns are both in the order by clause, which is the closest thing we have to a primary key, along with `timestamp`.
- `timestamp` and `retention_days` are in the partition key, so they need to remain, otherwise the tombstone record would end up in a different partition and never be merged.

It seems to me that it follows that these fields also need to be treated as generally immutable.

`deleted` and `group_id` don't have these same constraints, and I assume these are in here because these fields are not nullable in the schema definition.

- `deleted` is a special case, since [the value isn't carried over from the original row in deletions](https://github.com/getsentry/snuba/blob/master/snuba/replacer.py#L197). Should it be in here at all? It seems like it'd be more effective for this column to be defined separately.
- Should `group_id` be just be written to a static value like `0` that isn't part of the valid input domain? It doesn't seem necessary to keep the original value. I don't think we need to this now, but maybe worth considering. In general I think we should try and avoid preserving as many fields from the original row when deleting as possible.

Reply:
I picked these fields to preserve compatibility with the current deletion process. Essentially the deletion now is:
```
INSERT INTO sentry_dist (event_id, project_id, group_id, timestamp, deleted, retention_days)
SELECT event_id, project_id, group_id, timestamp, 1, retention_days
WHERE ....
```

I don't really know why group_id is there with the original value, but this is how replacer works now. https://github.com/getsentry/snuba/blob/master/snuba/replacer.py#L197

I am ok splitting required columns from primary key columns, but I am not sure about changing the behavior of group_id. 
About delete, I will actually remove it from here and let the replacer figure it out. Anyway it will be taken care of when I introduce prepared statements in step 3